### DISCUSSION TITLE
How to sign arbitrary message such as proof challenges using rust sdk ?

### DISCUSSION THREAD

**Discussion by aptgetgit** [aptos-labs/aptos-developer-discussions#59] [Questions]
Unable to figure out how to actually sign a rotation proof challenge, for verifying capability to rotate the authentication key.

My rotation proof challenge - 
```
let rotation_proof_challenge = RotationProofChallenge {
        account_address: PeerId::from_str("0x1").unwrap(),
        module_name: "account".to_string(),
        struct_name: "RotationProofChallenge".to_string(),
        sequence_number: alice_account_info.sequence_number,
        originator: alice_local_account.address(),
        current_auth_key: 
        PeerId::from_bytes( alice_local_account.public_key().to_bytes().to_vec())
            .unwrap(),
        new_public_key: multi_ed25519_public_key.to_bytes()
    };
```
    
need to pass the signed challenge to the aptos_framework_sdk_builder rotate function, [here](https://github.com/aptos-labs/aptos-core/blob/42d905a335a32002820253a993b4357b68a0a1de/aptos-move/framework/cached-packages/src/aptos_framework_sdk_builder.rs#L1703)
```
pub fn account_rotate_authentication_key(
    from_scheme: u8,
    from_public_key_bytes: Vec<u8>,
    to_scheme: u8,
    to_public_key_bytes: Vec<u8>,
    cap_rotate_key: Vec<u8>,
    cap_update_table: Vec<u8>,
) -> TransactionPayload {}
```

I am trying to use the sign_arbitrary_message exposed on the Ed25519PrivateKey struct [here](https://github.com/aptos-labs/aptos-core/blob/42d905a335a32002820253a993b4357b68a0a1de/crates/aptos-crypto/src/ed25519/ed25519_keys.rs#L60), but the function is private.

trying to find a workaround but seems like serialization is needed for the proof to be signed by other key types.

searching on discord, yields results for the typescript sdk  [here](https://discord.com/channels/945856774056083548/946123778793029683/1177041775047741450) but confused how to do it in rust
any help/guide/examples would be appreciated 
Thanks

**Comment by gregnazario** [comment]
The original authors of the crypto code were very protective against signing arbitrary data, but you can do one of two things:

1. Enable the `fuzzing` feature, and use this function from the trait https://github.com/aptos-labs/aptos-core/blob/42d905a335a32002820253a993b4357b68a0a1de/crates/aptos-crypto/src/ed25519/ed25519_keys.rs#L152-L155
2. Implement `CryptoHash` for the struct you want and then use this function. https://github.com/aptos-labs/aptos-core/blob/42d905a335a32002820253a993b4357b68a0a1de/crates/aptos-crypto/src/ed25519/ed25519_keys.rs#L142-L150

Here's an example of how key rotation is handled in the Aptos CLI https://github.com/aptos-labs/aptos-core/blob/main/crates/aptos/src/account/key_rotation.rs#L118-L136

**Reply by aptgetgit** [reply]
Hey @gregnazario , thanks for the reply.
Unable to do both, I might be missing something but here's what i did -

 - I followed the docs on how to install the aptos rust sdk - https://aptos.dev/sdks/rust-sdk/
 - Enabling the fuzzing feature throws error -
	 - Updated my `cargo.toml` file to this - 
		 - `aptos-sdk = { git = "https://github.com/aptos-labs/aptos-core",  branch = "devnet",  features = ['fuzzing']}`
	- which throws error -
		 - ```the package `my-package` depends on `aptos-sdk`, with features: `fuzzing` but `aptos-sdk` does not have these features.```
- For implementing  `CryptoHash` and `BCSCryptoHash`need to use the package `aptos-crypto-derive` which is not exposed for `aptos-sdk` installation i followed.

I think the `aptos-sdk` installation steps are wrong , but the crates were yanked from crates.io and the aptos docs suggests this way, so not sure what else to do.
